syntax = "proto3";

package kvdb;

// KVDB gRPC 服务定义
service KVDBService {
    // 基本操作
    rpc Put(PutRequest) returns (PutResponse);
    rpc Get(GetRequest) returns (GetResponse);
    rpc Delete(DeleteRequest) returns (DeleteResponse);
    
    // 批量操作
    rpc BatchPut(BatchPutRequest) returns (BatchPutResponse);
    rpc BatchGet(BatchGetRequest) returns (BatchGetResponse);
    
    // 扫描操作
    rpc Scan(ScanRequest) returns (stream ScanResponse);
    rpc PrefixScan(PrefixScanRequest) returns (stream PrefixScanResponse);
    
    // 快照操作
    rpc CreateSnapshot(CreateSnapshotRequest) returns (CreateSnapshotResponse);
    rpc ReleaseSnapshot(ReleaseSnapshotRequest) returns (ReleaseSnapshotResponse);
    rpc GetAtSnapshot(GetAtSnapshotRequest) returns (GetAtSnapshotResponse);
    
    // 管理操作
    rpc Flush(FlushRequest) returns (FlushResponse);
    rpc Compact(CompactRequest) returns (CompactResponse);
    rpc GetStats(GetStatsRequest) returns (GetStatsResponse);
    rpc SetCompactionStrategy(SetCompactionStrategyRequest) returns (SetCompactionStrategyResponse);
    
    // 实时订阅
    rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse);
}

// 基本操作消息
message PutRequest {
    string key = 1;
    string value = 2;
}

message PutResponse {
    bool success = 1;
    string error_message = 2;
}

message GetRequest {
    string key = 1;
}

message GetResponse {
    bool found = 1;
    string value = 2;
    string error_message = 3;
}

message DeleteRequest {
    string key = 1;
}

message DeleteResponse {
    bool success = 1;
    string error_message = 2;
}

// 批量操作消息
message KeyValue {
    string key = 1;
    string value = 2;
}

message BatchPutRequest {
    repeated KeyValue pairs = 1;
}

message BatchPutResponse {
    bool success = 1;
    int32 processed_count = 2;
    string error_message = 3;
}

message BatchGetRequest {
    repeated string keys = 1;
}

message BatchGetResponse {
    repeated KeyValue pairs = 1;
    string error_message = 2;
}

// 扫描操作消息
message ScanRequest {
    string start_key = 1;
    string end_key = 2;
    int32 limit = 3;
}

message ScanResponse {
    string key = 1;
    string value = 2;
}

message PrefixScanRequest {
    string prefix = 1;
    int32 limit = 2;
}

message PrefixScanResponse {
    string key = 1;
    string value = 2;
}

// 快照操作消息
message CreateSnapshotRequest {
}

message CreateSnapshotResponse {
    uint64 snapshot_id = 1;
    string error_message = 2;
}

message ReleaseSnapshotRequest {
    uint64 snapshot_id = 1;
}

message ReleaseSnapshotResponse {
    bool success = 1;
    string error_message = 2;
}

message GetAtSnapshotRequest {
    string key = 1;
    uint64 snapshot_id = 2;
}

message GetAtSnapshotResponse {
    bool found = 1;
    string value = 2;
    string error_message = 3;
}

// 管理操作消息
message FlushRequest {
}

message FlushResponse {
    bool success = 1;
    string error_message = 2;
}

message CompactRequest {
}

message CompactResponse {
    bool success = 1;
    string error_message = 2;
}

message GetStatsRequest {
}

message GetStatsResponse {
    uint64 memtable_size = 1;
    uint64 wal_size = 2;
    double cache_hit_rate = 3;
    int32 active_snapshots = 4;
    CompactionStats compaction_stats = 5;
}

message CompactionStats {
    uint64 total_compactions = 1;
    uint64 bytes_read = 2;
    uint64 bytes_written = 3;
    double write_amplification = 4;
    uint64 total_time_ms = 5;
}

enum CompactionStrategyType {
    LEVELED = 0;
    TIERED = 1;
    SIZE_TIERED = 2;
    TIME_WINDOW = 3;
}

message SetCompactionStrategyRequest {
    CompactionStrategyType strategy = 1;
}

message SetCompactionStrategyResponse {
    bool success = 1;
    string error_message = 2;
}

// 实时订阅消息
message SubscribeRequest {
    string key_pattern = 1;  // 支持通配符
    bool include_deletes = 2;
}

message SubscribeResponse {
    string key = 1;
    string value = 2;
    string operation = 3;  // PUT, DELETE
    uint64 timestamp = 4;
}